File: src/layoutComponents.ts
export const layoutComponents = (
  board: number[][],
  components: number[][][],
  absoluteComponents: { characters: number[][]; x: number; y: number }[],
  calendarComponents?: { characters: number[][]; x: number }[]
) => {
  let position = {
    top: 0,
    left: 0,
    height: 0,
  };

  components.forEach((component) => {
    // If the component size plus the currently occupied size is larger than the board width, flow to the next line
    const newLine = position.left + component[0].length > board[0].length;
    const left = newLine ? 0 : position.left;
    const top = newLine ? position.top + position.height : position.top;

    // Fill in the individual component bits over the empty board
    component.forEach((row, rowIndex) => {
      row.forEach((bit, bitIndex) => {
        board[rowIndex + top][bitIndex + left] = bit;
      });
    });

    // Update the last position of this component for reference on the next iteration
    position = {
      top: top,
      left: left + component[0].length,
      height: component.length,
    };
  });

  absoluteComponents &&
    absoluteComponents.forEach((component) => {
      component.characters.forEach((row, rowIndex) => {
        row.forEach((bit, bitIndex) => {
          // make sure we are in bounds, truncate the rest for now
          if (component.y + rowIndex >= board.length) {
            return;
          }
          if (component.x + bitIndex >= board[0].length) {
            return;
          }
          board[rowIndex + component.y][bitIndex + component.x] = bit;
        });
      });
    });

  calendarComponents &&
    calendarComponents.forEach((component) => {
      component.characters.forEach((row, rowIndex) => {
        row.forEach((bit, bitIndex) => {
          // make sure we are in bounds, truncate the rest for now
          if (rowIndex >= board.length) {
            return;
          }
          // calendars are always 12 wide
          if (
            component.x + bitIndex >= board[0].length ||
            component.x + bitIndex > 12
          ) {
            return;
          }
          board[rowIndex][bitIndex + component.x] = bit;
        });
      });
    });

  return board;
};

File: python/pyvbml/layout_components.py
"""Layout components.

Port of Vestaboard/vbml/src/layoutComponents.ts
"""

from __future__ import annotations


def layout_components(
    board: list[list[int]],
    components: list[list[list[int]]],
    absolute_components: list[dict],
    calendar_components: list[dict] | None = None,
) -> list[list[int]]:
    """Layout components."""
    position = {"top": 0, "left": 0, "height": 0}
    component: list[list[int]] | dict

    for component in components:
        # If the component would overflow the board width, wrap to the next line
        new_line = position["left"] + len(component[0]) > len(board[0])
        left = 0 if new_line else position["left"]
        top = (position["top"] + position["height"]) if new_line else position["top"]

        for row_idx, row in enumerate(component):
            for bit_idx, bit in enumerate(row):
                if row_idx + top >= len(board):
                    continue
                if bit_idx + left >= len(board[0]):
                    continue
                board[row_idx + top][bit_idx + left] = bit

        position = {
            "top": top,
            "left": left + len(component[0]),
            "height": len(component),
        }

    for component in absolute_components or []:
        for row_idx, row in enumerate(component["characters"]):
            for bit_idx, bit in enumerate(row):
                # Out-of-bounds: truncate
                if component["y"] + row_idx >= len(board):
                    continue
                if component["x"] + bit_idx >= len(board[0]):
                    continue
                board[row_idx + component["y"]][bit_idx + component["x"]] = bit

    for component in calendar_components or []:
        for row_idx, row in enumerate(component["characters"]):
            for bit_idx, bit in enumerate(row):
                if row_idx >= len(board):
                    continue
                # Calendars are always 12 wide
                if component["x"] + bit_idx >= len(board[0]) or bit_idx > 12:
                    continue
                board[row_idx][bit_idx + component["x"]] = bit

    return board
